### ISSUE TITLE
Export Facebook to Google Calendar

### ISSUE THREAD

**Issue by uncovery** [pointofgravity/HKMGigAnnouncements#4]
Hi,
I have seen you created a project for this. There is a way to do this by simply exporting all your events automatically as a iCal file or by exporting a single event.

When you open any event, there is the [...] icon on the top right. In that menu, there is a "Export Event".
From here, there are 2 options: 

1. Save to calendar: This will create an ics file that could be inserted into the google calendar or the application we are creating here.
2. "Subscribe to all upcoming events on your calendar": This only works if there are no public (non-music) events in the personal calendar, otherwise those would be exported as well. Since public events are marked with "CLASS:PUBLIC", we can parse the events for that and only add those to the google calendar. 

I tried to auto-download the file via PHP and so far failed. we might need to download it manually each time and upload it to the system.

**Comment by pointofgravity** [comment]
Hi, thanks for taking time to looking into this, I've been really busy with work recently so I haven't been able to myself.

This is a good feature we can use, we can use this in two ways:

1) Monitor a page that shares other people's events, like the Hong Kong Gig Guide
Or
2) Monitor a page or pages that exclusively publish their events, like Underground HK and The Wanch, and various band pages/ events companies.

Since these pages' motivation is to promote their show, there is no danger of sharing private events. If the events were private, that would defeat the purpose.

If we integrate this feature, our course of action would be:

1) Create an algorithm to monitor the page for events
2) Create a function to export the event as an iCal event
3) Create an algorithm to search for keywords like "Doors", "Price", "Tickets" etc.
4) Create a function to parse date, Time, Location and title text
5) Create a function to concat this information into strings
6) Create a function that will create a Google calendar event using the data and time data (php?)
7) Create a function that will import the strings into the Title, Date and description fields of the Google Calendar field.
8) Save the event

IFTTT will handle the rest. If we get this underway, we can K.O. the project in one.

I will try to take a look at these tasks, but I am very new to using APIs and have only written one php script before, so I dont expect I will be able to hack it very quickly. On top of that my deadlines at work have piled up, I should be more free after January.

Are you in Hong Kong? Maybe we can grab a bite sometime.

Thanks,
Pointofgravity

Sent from my Huawei Mobile

**Comment by uncovery** [comment]
What you propose on the top is tricky. Whats your proposal to monitor and then and then export them automatically?

The topic regarding "private" events is a misunderstanding. I meant:
If you want to use the webcal:// format file, it will automatically contain all the events you personally subscribed to. If you auto-import those to the target calendar, your other, personal, private events are in the same file.

the top 2 monitor parts are the hardest. once we have an ical file, the rest is easy in PHP.

Yes, I am in HKG, sure we can have a drink or the like sometime.

**Comment by pointofgravity** [comment]
We could check the page for events posted in the last x hours, that's how the sidebar updates.

Sent from my Huawei Mobile

-------- Original Message --------
Subject: Re: [pointofgravity/HKMGigAnnouncements] Export Facebook to Google Calendar (#4)
From: Uncovery
To: pointofgravity/HKMGigAnnouncements
CC: pointofgravity ,Comment

What you propose on the top is tricky. Whats your proposal to monitor and then and then export them automatically?

The topic regarding "private" events is a misunderstanding. I meant:
If you want to use the webcal:// format file, it will automatically contain all the events you personally subscribed to. If you auto-import those to the target calendar, your other, personal, private events are in the same file.

the top 2 monitor parts are the hardest. once we have an ical file, the rest is easy in PHP.

Yes, I am in HKG, sure we can have a drink or the like sometime.

**Comment by pointofgravity** [comment]
Hey, 

I've had some time to check your options out, it seems like the first option is the best one.  Now we just need a way to check a facebook page for events published in the last x hours. I have added the google script that I use for updating the subreddit sidebar (HKM_Reddit_Sidebar.gs) to the repo, it uses a similar mechanic to check for new events from the google calendar every four hours. Perhaps we can use similar code from that.

**Comment by uncovery** [comment]
So I found out more about how the mechanics of the facebook events work. Unfortunately, there does not seem to be an easy way to automate stuff 100%.

-  I tried to subscribe to event hosts (pages) and those I cannot see in the file that I can export at all. One has to "Join" the events for them to be visible in the file
- Once I download the file and add it to google calendar, it does not seem to update if I add things later.
- I tried to download the events section from a facebook page. That does not seem to work since you need a browser to load the JS/Ajax loaded content. I tried to use php/CURL with SSL, I get a page, but the events section is empty. 

So from my perspective, it's not possible to automatically get events from facebook. The most automated solution would be to manually copy/paste/save the HTML of every single event from facebook into a parser and then have it automatically transferred to a database/ical.

However, I do not think it's realistic to permanently do that, week after week.

**Comment by pointofgravity** [comment]
Hey man,

Don't worry about the Facebook events thing then, I think it's not a priority. We can do something a bit easier, like using the email from the underground.

Sent from my Huawei Mobile


-------- Original Message --------
Subject: Re: [pointofgravity/HKMGigAnnouncements] Export Facebook to Google Calendar (#4)
From: Uncovery
To: pointofgravity/HKMGigAnnouncements
CC: pointofgravity ,Comment



So I found out more about how the mechanics of the facebook events work. Unfortunately, there does not seem to be an easy way to automate stuff 100%.

  *   I tried to subscribe to event hosts (pages) and those I cannot see in the file that I can export at all. One has to "Join" the events for them to be visible in the file
  *   Once I download the file and add it to google calendar, it does not seem to update if I add things later.
  *   I tried to download the events section from a facebook page. That does not seem to work since you need a browser to load the JS/Ajax loaded content. I tried to use php/CURL with SSL, I get a page, but the events section is empty.

So from my perspective, it's not possible to automatically get events from facebook. The most automated solution would be to manually copy/paste/save the HTML of every single event from facebook into a parser and then have it automatically transferred to a database/ical.

However, I do not think it's realistic to permanently do that, week after week.

**Comment by uncovery** [comment]
Yeah it's just really a pity since more and more bands a creating FB events and it would be a common format to work with. The newsletter from the underground can look different each time and it's manual work again.

**Comment by pointofgravity** [comment]
Do you want to go through this over a beer some time? I think we need to rethink the whole thing.b

Sent from my Huawei Mobile


-------- Original Message --------
Subject: Re: [pointofgravity/HKMGigAnnouncements] Export Facebook to Google Calendar (#4)
From: Uncovery
To: pointofgravity/HKMGigAnnouncements
CC: pointofgravity ,Comment



Yeah it's just really a pity since more and more bands a creating FB events and it would be a common format to work with. The newsletter from the underground can look different each time and it's manual work again.

**Comment by uncovery** [comment]
Yeah that sounds like a good idea. I am right now a bit on a dead end with this. Either we find a completely new way how to automate this stuff or we drop the project as a whole. Another angle is that I do not speak cantonese and I suspect that there is a whole huge amount of info in Cantonese out there that I don't even know about. I'll send you a reddit PM.

**Comment by pointofgravity** [comment]
Hey dude, are you here? Can't get a hold of you man.

Sent from my Huawei Mobile


-------- Original Message --------
Subject: Re: [pointofgravity/HKMGigAnnouncements] Export Facebook to Google Calendar (#4)
From: Uncovery
To: pointofgravity/HKMGigAnnouncements
CC: pointofgravity ,Comment



Yeah that sounds like a good idea. I am right now a bit on a dead end with this. Either we find a completely new way how to automate this stuff or we drop the project as a whole. Another angle is that I do not speak cantonese and I suspect that there is a whole huge amount of info in Cantonese out there that I don't even know about. I'll send you a reddit PM.

**Comment by pointofgravity** [comment]
Hi man,

Hope you're doing well. I think I have found a way to automate this. The downside is that we are only going to be able to automate it from one source, which is the The Underground newsletter. 

The Underground newsletter's gig listing always begins under the header:
未來2星期的現場音樂演出   Live Music Shows Happening In The Next Two Weeks

![screen shot 2017-06-03 at 2 04 36 pm](https://cloud.githubusercontent.com/assets/10907544/26751174/a2d92e64-4865-11e7-88da-36a4eac5c389.png)

If we could subscribe this newsletter to an e-mail account, could we parse the information on there? From then on, the format is pretty consistent: Date, Location, Time, Bands and Price. What I think is handy still is that there are clear markers for these pieces of information. I am still insistent on using the Google Calendar format; Since taking ownership of the Hong Kong Gig Guide, I have changed from Facebook Notes format to using this format to display gig dates, it is convenient and users can see live changes better. 

I have found an integration tool called Zapier that can allow me to create a google sheet, define columns that are input into different text box fields in a google calendar:
![screen shot 2017-06-03 at 2 44 43 pm](https://cloud.githubusercontent.com/assets/10907544/26751402/37976f66-486b-11e7-81c4-d82eae025c8c.png)

Using this method, we can finally automate updating the Hong Kong Gig Guide calendar. It is a shame that smaller events not listed in The Underground's newsletter will not be included, but I think that the newsletter is pretty comprehensive already. The only issue I can think of is that if I already added an event before (for example, Clockenflap is announced and I input it into the calendar) and then the newsletter gets released (it is out every two weeks and only includes events happening in the next two weeks) then when the zapier app updates, we will get two of the same events, so it may need some manual maintenance after all. 

So in conclusion, I think the only thing I need now is parsing the information from the newsletter and then entering it into this spreadsheet: https://docs.google.com/spreadsheets/d/1iauJpFSXXk9NU2yM1Mt87dw1AWPtOeLwef25Is6fVy8/edit?usp=sharing 
Then, the zapier app will take care of the rest. I also have an issue further on about how I currently have a bot posting to the subreddit every time an event is added to the calendar, but I think I will tackle that later on, I think this should take more priority for now.

**Comment by uncovery** [comment]
Did you look at this here? https://parser.zapier.com/ This actually takes emails as a data source for zapier.

**Comment by uncovery** [comment]
by the way, are you aware about this page? http://timable.com/music

**Comment by pointofgravity** [comment]
Hi, yes sorry I didn't reply for so long. I'll look in to the parser zap, thanks for the pointer, and yes I'm aware of the timable calendar.
